src/OpenTelemetry/Internal/ProviderBuilderState.cs

```diff
@@ -0,0 +1,94 @@
+// <copyright file="ProviderBuilderState.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using OpenTelemetry.Resources;
+
+namespace OpenTelemetry;
+
+internal abstract class ProviderBuilderState<TBuilder, TProvider>
+{
+    protected ProviderBuilderState(
+        IServiceProvider serviceProvider,
+        TProvider provider)
+    {
+        Debug.Assert(serviceProvider != null, "serviceProvider was null");
+        Debug.Assert(provider != null, "provider was null");
+
+        this.ServiceProvider = serviceProvider!;
+        this.Provider = provider;
+    }
+
+    public IServiceProvider ServiceProvider { get; }
+
+    public abstract TBuilder Builder { get; }
+
+    public TProvider Provider { get; }
+
+    public List<InstrumentationRegistration> Instrumentation { get; } = new();
+
+    public ResourceBuilder? ResourceBuilder { get; protected set; }
+
+    public void AddInstrumentation(
+        string instrumentationName,
+        string instrumentationVersion,
+        object instrumentation)
+    {
+        Debug.Assert(!string.IsNullOrWhiteSpace(instrumentationName), "instrumentationName was null or whitespace");
+        Debug.Assert(!string.IsNullOrWhiteSpace(instrumentationVersion), "instrumentationVersion was null or whitespace");
+        Debug.Assert(instrumentation != null, "instrumentation was null");
+
+        this.Instrumentation.Add(
+            new InstrumentationRegistration(
+                instrumentationName,
+                instrumentationVersion,
+                instrumentation!));
+    }
+
+    public void ConfigureResource(Action<ResourceBuilder> configure)
+    {
+        Debug.Assert(configure != null, "configure was null");
+
+        var resourceBuilder = this.ResourceBuilder ??= ResourceBuilder.CreateDefault();
+
+        configure!(resourceBuilder);
+    }
+
+    public void SetResourceBuilder(ResourceBuilder resourceBuilder)
```

Comment:
I personally like the `ConfigureResource` pattern over this. Does it make sense to just remove it from logging, or do you think the lack of parity with our trace/metric support is bad?

Reply:
I'm not totally convinced on the uselessness of `SetResourceBuilder`. If you look at any individual signal in a vacuum, it totally makes sense to use `ConfigureResource`. But if you are configuring a shared/global resource which will be used by two or more signals, or two or more providers of the same signal, isn't `SetResourceBuilder` more useful?

Reply:
Yea good point. It may still be preferred by some. I've been shoving resource configuration into a method then using it like `providerBuilder.ConfigureResource(ConfigureMyResource)`.